src/classlibnative/bcltype/bignum.cpp

```diff
@@ -0,0 +1,597 @@
+{
+    BigNum temp;
+    BigNum::Multiply(*this, value, temp);
+
+    memcpy(m_blocks, temp.m_blocks, ((UINT32)temp.m_len) * sizeof(UINT32));
+    m_len = temp.m_len;
+}
+
+void BigNum::Multiply(const BigNum& lhs, UINT32 value, BigNum& result)
+{
+    if (lhs.m_len == 0)
+    {
+        return;
+    }
+
+    if (lhs.m_len < BIGSIZE)
+    {
+        // Set the highest + 1 bit to zero so that
+        // we can check if there's a carry later.
+        result.m_blocks[lhs.m_len] = 0;
+    }
+
+    const UINT32* pCurrent = lhs.m_blocks;
+    const UINT32* pEnd = pCurrent + lhs.m_len;
+    UINT32* pResultCurrent = result.m_blocks;
+
+    UINT64 carry = 0;
+    while (pCurrent != pEnd)
+    {
+        UINT64 product = (UINT64)(*pCurrent) * (UINT64)value + carry;
+        carry = product >> 32;
+        *pResultCurrent = (UINT32)(product & 0xFFFFFFFF);
+
+        ++pResultCurrent;
+        ++pCurrent;
+    }
+
+    if (lhs.m_len < BIGSIZE && result.m_blocks[lhs.m_len] != 0)
+    {
+        result.m_len = lhs.m_len + 1;
+    }
+    else
+    {
+        result.m_len = lhs.m_len;
+    }
+}
+
+void BigNum::Multiply(const BigNum& lhs, const BigNum& rhs, BigNum& result)
+{
+    const BigNum* pLarge = NULL;
+    const BigNum* pSmall = NULL;
+    if (lhs.m_len < rhs.m_len)
+    {
+        pSmall = &lhs;
+        pLarge = &rhs;
+    }
+    else
+    {
+        pSmall = &rhs;
+        pLarge = &lhs;
+    }
+
+    UINT8 maxResultLength = pSmall->m_len + pLarge->m_len;
+
+    // Zero out result internal blocks.
+    memset(result.m_blocks, 0, sizeof(UINT32) * BIGSIZE);
+
+    const UINT32* pLargeBegin = pLarge->m_blocks;
+    const UINT32* pLargeEnd = pLarge->m_blocks + pLarge->m_len;
+
+    UINT32* pResultStart = result.m_blocks;
+    const UINT32* pSmallCurrent = pSmall->m_blocks;
+    const UINT32* pSmallEnd = pSmallCurrent + pSmall->m_len;
+
+    while (pSmallCurrent != pSmallEnd)
+    {
+        // Multiply each block of large BigNum.
+        if (*pSmallCurrent != 0)
+        {
+            const UINT32* pLargeCurrent = pLargeBegin;
+            UINT32* pResultCurrent = pResultStart;
+            UINT64 carry = 0;
+
+            do
+            {
+                UINT64 product = (UINT64)(*pResultCurrent) + (UINT64)(*pSmallCurrent) * (UINT64)(*pLargeCurrent) + carry;
+                carry = product >> 32;
+                *pResultCurrent = (UINT32)(product & 0xFFFFFFFF);
+
+                ++pResultCurrent;
+                ++pLargeCurrent;
+            } while (pLargeCurrent != pLargeEnd);
+
+            *pResultCurrent = (UINT32)(carry & 0xFFFFFFFF);
+        }
+
+        ++pSmallCurrent;
+        ++pResultStart;
+    }
+
+    if (maxResultLength > 0 && result.m_blocks[maxResultLength - 1] == 0)
+    {
+        result.m_len = maxResultLength - 1;
+    }
+    else
+    {
+        result.m_len = maxResultLength;
+    }
+}
+
+bool BigNum::IsZero()
+{
+    if (m_len == 0)
+    {
+        return true;
+    }
+
+    for (UINT8 i = 0; i < m_len; ++i)
```

Comment:
IsZero is part of the `BigNum` class, so it would only be dependent on a convention set for its own class.

This probably isn't too important so we can investigate later.